Wip dash

rzerres opened the pull request:
Hey Wes,

since i needed a working btrfs based backup at customer site, i was digging the net and learned about the available solutions. So i also got hold of snap-sync projekt.

I loved to see, that you are synchronizing based on snapper administered subvolumes. That is how i wanted to keep track on the source as well. I'm aiming to realize a disk-to-disk-to-tape solution, using snapper as far as possible. As a requirement, this can only work out, if all involved filesystems will be btrfs based. Kernels are mostly v4.11, which will benefit from new btrfs-progs bugfixes as well. All fine until there.

Like lots of other smaller sites, one big productions system is used a host for VM's and containers.
There is the need to lower downtime and complexity. One way to go is to have a 2nd server, lots of storage space in a different floor (network is vlan tagged with 10Gbit, switches are linked with fiber).
Snap-sync could backup from source to target right now. But it will not let me save in separate/specific mounted subvolumes on the target. Until yet is suggested to have subfolders (backupdir) on a unice addressable disk/disk-group  (UUID).

Since i love the ability to use subvolumes/snapshots, i thought i give it a try and implement a structure, where you can assing a dedicated target subvolume (subvolid) as destination of the btrfs-send/btrfs-recieve pipe. This can be tracked in the userdata key/value pairs of snapper's config log. So i extended the code as a first try.

This worked out.

Now i got a bit deeper and wanted get smarter in  the selection dialog. If you don't select a disk at all, present the list of all available disks. If you know the UUID of the disk, only show possible subvolumes on this disk. If you know the subvolid (the mountpoint) you can adress it directly. This is also implemented. The dialogs will present the extended information needed.

To keep the logic, that you can trigger specific or all snapper configs as sources of sync jobs, each config need to keep track, where the data will be synced to. Combining UUID and SUBVOLID solves this problem to have a unique backup-root. And you can keep the possibility to define a backup-dir as a subfolder-structure on this backup-root. 

Since bashism was a no go here, i took the chance to convert the bash script to a posix comlient one. I tested with dash. The Biggest problem was  the absense of array functionality, since it is not defined in the posix standard. I came up to use some king of "pseudo array" logic. Have a look in the code.

I have a working version in my WIP branch. I bumped the release to version 0.4.4. If you have time to check it out, it would be a pleasure for me. There is more to come up with. But this is the intro.

best regards
Ralf

rzerres committed:
README.md: adapted the documentation

 - aknowledge posix shift
 - update help section
 - update example section to be in line with new code

rzerres committed:
Git: ignore reproducable contents

rzerres committed:
TODO.md: scheduler for open tasks

- structure ideas on how to improve the wrapper
- keep track on programming tasks

rzerres committed:
snap-sync.bash: rename bash version of snap-sync to snap-sync.bash

rzerres committed:
snap-sync: new working version

rzerres committed:
snap-sync: change progname variable

- update the variable to hold the progname
- eleminate trailing whitespace

rzerres committed:
snap-sync: new function parse_params()

- create a parse function (check initial arguments and defaults)
- introduce a verbose option

rzerres committed:
snap-sync: new function check_prerequisites()

- check availability of referenced programs at runtime
- check if $SNAPPER_CONFIG is readable

rzerres committed:
snap-sync: restructure code to use logical function blocks

- new function get_disk_info()
- new function run_config()
- new function run_backup()
- new function get_target_disk()
- apply them in logical order

rzerres committed:
snap-sync: inital posix compliant version

- use dash as reference shell
- remove all bashism
- convert bash arrays to posix "pseudo arrays"
- introduce global/local variables
- using (hopefully) more descriptive variables
- new function verify_snapper_structure()
  that takes care to construct target snapper file-structure
  as needed

rzerres committed:
snap-sync: support dryrun, and selection of mounted target subvolume

- introduce commandline option --dry-run
  perform a trial run where no changes are made.
- introduce option --TARGET
  Specify the mountpoint of the BTRFS subvolume to back up to.
  This makes it possible, to have multiple subvolumes on target disks,
  and select them as needed to store backup configurations.
- some cleanups

rzerres committed:
snap-sync: bump version to 0.4.4

wesbarnett commented:
Wow, these are a lot of ideas! I've looked through it some, but this is a lot of changes so I'm going to have to take some time with it.

wesbarnett commented on bin/snap-sync.bash:
```diff
@@ -70,33 +70,33 @@ traperror() {
                           configuration. Can list multiple configurations within quotes, space-separated
                           (e.g. -c "root home").
  -n, --noconfirm          Do not ask for confirmation for each configuration. Will still prompt for backup
-                          directory name on first backup"
```
It looks like you did a search and replace which changes "name" to progname". Some of these should not have been changed.

rzerres committed:
snap-sync: reflact subvolid for processed snapshot bakcup

- snapper_sync_id should take into account the selected_subvolid
  in combination with the selected_uuid

rzerres commented:
Will rebase wip-dash to reflect your changes up to 0875b5e.
this will ease follow-ups ...
